src/Tools/dotnet-monitor/StartupHook/StartupHookEndpointInfoSourceCallback.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Diagnostics.Monitoring.WebApi;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Tools.Monitor.StartupHook
+{
+    internal sealed class StartupHookEndpointInfoSourceCallbacks : IEndpointInfoSourceCallbacks
+    {
+        private readonly IInProcessFeatures _inProcessFeatures;
+        private readonly StartupHookValidator _startupHookValidator;
+
+
+        public StartupHookEndpointInfoSourceCallbacks(
+            IInProcessFeatures inProcessFeatures,
+            StartupHookValidator startupHookValidator)
+        {
+            _inProcessFeatures = inProcessFeatures;
+            _startupHookValidator = startupHookValidator;
+        }
+
+        Task IEndpointInfoSourceCallbacks.OnAddedEndpointInfoAsync(IEndpointInfo endpointInfo, CancellationToken cancellationToken)
+        {
+            return Task.CompletedTask;
+        }
+
+        Task IEndpointInfoSourceCallbacks.OnBeforeResumeAsync(IEndpointInfo endpointInfo, CancellationToken cancellationToken)
+        {
+            if (_inProcessFeatures.IsStartupHookRequired)
+            {
+                if (_startupHookValidator.CheckEnvironmentAsync(endpointInfo, cancellationToken).Result)
```

Comment:
Please change OnBeforeResume to be `async` and await this method instead of using `.Result`.